File: native/tools/dump_solid.cpp
// Stage 2 verification, geometry half.
//
// Builds every room in the folder and prints the numbers the Python build is
// already trusted for: solid, shell and face counts, volume, and the wall
// tiling check where the exported walls have to add back up to the room.
#include <cstdio>
#include <iostream>
#include <string>

#include "snapir/parser.hpp"
#include "snapir/solid.hpp"
#include "snapir/tessellate.hpp"

using namespace snapir;

namespace {

void emit(const std::string& room, const std::string& key, const std::string& value) {
  std::cout << room << '|' << key << '|' << value << '\n';
}

std::string num(double v, int places = 9) {
  char buf[64];
  std::snprintf(buf, sizeof(buf), "%.*f", places, v);
  return buf;
}

}  // namespace

int main(int argc, char** argv) {
  if (argc < 2) {
    std::cerr << "usage: dump_solid <folder> [--fmt step|iges|brep <out_dir>]\n";
    return 2;
  }

  // Optional: also write every room out, so one command produces the files a
  // format change has to be judged on.
  std::string fmt, out_dir;
  for (int i = 2; i + 1 < argc; ++i)
    if (std::string(argv[i]) == "--fmt") {
      fmt = argv[i + 1];
      out_dir = i + 2 < argc ? argv[i + 2] : ".";
    }
  if (!fmt.empty() && !is_export_format(fmt)) {
    std::cerr << "dump_solid: unknown format " << fmt << "\n";
    return 2;
  }

  const BuildSettings cfg;
  Project proj = read_project(argv[1]);

  for (auto& room : proj.rooms) {
    const std::string& n = room.name;
    TopoDS_Shape shape;
    try {
      shape = build_room(room, cfg);
    } catch (const std::exception& e) {
      emit(n, "build", std::string("ERROR ") + e.what());
      continue;
    }

    const SolidStats s = solid_stats(shape);
    emit(n, "build", "ok");
    emit(n, "solids", std::to_string(s.solids));
    emit(n, "shells", std::to_string(s.shells));
    emit(n, "faces", std::to_string(s.faces));
    emit(n, "volume_m3", num(s.volume_m3));

    if (!fmt.empty()) {
      try {
        emit(n, "wrote", export_shape(shape, out_dir + "/" + n, fmt));
      } catch (const std::exception& e2) {
        emit(n, "wrote", std::string("ERROR ") + e2.what());
      }
    }

    const Mesh mesh = tessellate(shape);
    emit(n, "triangles", std::to_string(mesh.triangle_count()));
    emit(n, "mesh_faces", std::to_string(mesh.faces.size()));

    // The wall tiling property: every wall of the room, cut out on its own,
    // has to add back up to the room's own volume.
    double total = 0.0;
    bool all_ok = true;
    for (size_t e = 0; e < room.outline.size(); ++e) {
      try {
        const WallBody w = wall_body(room, cfg, static_cast<int>(e));
        const SolidStats ws = solid_stats(w.shape);
        total += ws.volume_m3;
        emit(n, "wall[" + std::to_string(e) + "]",
             "len=" + num(w.length, 6) + " solids=" + std::to_string(w.solids) +
                 " vol=" + num(ws.volume_m3));
      } catch (const std::exception& e2) {
        all_ok = false;
        emit(n, "wall[" + std::to_string(e) + "]", std::string("ERROR ") + e2.what());
      }
    }
    if (all_ok) emit(n, "wall_sum_m3", num(total));
  }
  return 0;
}

File: tools/dump_solid.py
"""Reference numbers for the C++ port, geometry half.

Builds every room and prints the numbers the Python build is already trusted
for. native/tools/dump_solid.cpp prints the same lines.
"""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

# The C++ side writes UTF-8. Windows hands Python a cp1252 stdout by default,
# which cannot encode a Turkish room name and kills the dump before it prints
# anything - so the two halves could not be compared at all on the machine the
# port is developed on.
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")


from snapir.parser import read_project
from snapir.settings import BuildSettings
from snapir.solid import build_room, solid_stats, wall_body
from snapir.tessellate import tessellate


def num(v: float, places: int = 9) -> str:
    return f"{v:.{places}f}"


def main(folder: str) -> None:
    cfg = BuildSettings()
    out: list[str] = []

    for room in read_project(folder).rooms:
        n = room.name

        def emit(key, value):
            out.append(f"{n}|{key}|{value}")

        try:
            shape = build_room(room, cfg)
        except Exception as exc:  # noqa: BLE001 - the message is the datum
            emit("build", f"ERROR {exc}")
            continue

        s = solid_stats(shape)
        emit("build", "ok")
        emit("solids", s["solids"])
        emit("shells", s["shells"])
        emit("faces", s["faces"])
        emit("volume_m3", num(s["volume_m3"]))

        mesh = tessellate(shape)
        emit("triangles", mesh.triangle_count)
        emit("mesh_faces", len(mesh.faces))

        total = 0.0
        all_ok = True
        for e in range(len(room.outline)):
            try:
                body, length, count = wall_body(room, cfg, e)
                ws = solid_stats(body)
                total += ws["volume_m3"]
                emit(f"wall[{e}]",
                     f"len={num(length, 6)} solids={count} vol={num(ws['volume_m3'])}")
            except Exception as exc:  # noqa: BLE001
                all_ok = False
                emit(f"wall[{e}]", f"ERROR {exc}")
        if all_ok:
            emit("wall_sum_m3", num(total))

    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main(sys.argv[1])
